src/client/Client.cc

```diff
@@ -4833,7 +4836,8 @@ void Client::trim_caps(MetaSession *s, uint64_t max)
       }
       if (all && in->ino != CEPH_INO_ROOT) {
         ldout(cct, 20) << __func__ << " counting as trimmed: " << *in << dendl;
-	trimmed++;
```

Comment:
Umm.. I thought the trimming for such inodes would be skipped altogether.

Reply:
I don't think so. Here we can just try to trim the dentries and then later after the dirty caps being flushed the clients should release the caps. This is also what kclient will do.

Reply:
I am thinking if we just skip trimming the `dentries` will be buggy ? Such as if after iterating all the inodes we still couldn't get enough inodes to release, and we just trigger to flush the dirty caps, then after the dirty caps triggered the clients couldn't release the caps by itself. And we must issue a new recall request to recall them.

Currently no matter there is any dirty caps we will always try to trim the dentries, and just after the dirty caps flushed the corresponding caps will be released.

Reply:
> I don't think so. Here we can just try to trim the dentries and then later after the dirty caps being flushed the clients should release the caps. This is also what kclient will do.

You are correct. The dentry can be trimmed and that would help with the client dropping states when the mds requests a state recall. The inode itself gets deallocated at a later point when its ref drops to 0.